src/Microsoft.Diagnostics.Monitoring.StartupHook/BackgroundService.cs

```diff
@@ -0,0 +1,50 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.StartupHook
+{
+    internal abstract class BackgroundService : IDisposable
+    {
+        private readonly CancellationTokenSource _cts = new();
+        private long _disposedState;
+
+        public Task? ExecutingTask { get; private set; }
+
+        public void Start()
+        {
+            ExecutingTask = Task.Run(async () =>
+            {
+                await ExecuteAsync(_cts.Token).ConfigureAwait(false);
+            }, _cts.Token);
+        }
+
+        public void Stop()
+        {
+            _cts.Cancel();
+
+            try
+            {
+                ExecutingTask?.Wait(TimeSpan.FromSeconds(1));
+            }
+            catch
+            {
+                // ignore
+            }
+        }
+
+        public virtual void Dispose()
+        {
+            if (!DisposableHelper.CanDispose(ref _disposedState))
+                return;
+
+            _cts.Cancel();
```

Comment:
Don't we also need to wrap `_cs.Cancel()` in a try/catch since it could throw? (same applies to `Stop`)

Reply:
It should only throw if `_cts` is disposed, which it should be since there is the disposal guard on the previous lines.

Reply:
`_cts.Cancel()` would throw if anything registers a delegate to be called when the corresponding token gets cancelled and it happens to throw.

e.g.

```csharp
  _cts.Token.Register(() =>
  {
      throw new InvalidOperationException("bug");
  });

  _cts.Cancel(); // <- throws 
```

Since we hand off `_cts`'s cancellation token to `abstract ExecuteAsync` it's possible that something like this could happen

Reply:
That's a good point. I think I should catch `AggregateException` (it wrapps any exception thrown from registered callbacks) and let `ObjectDisposedException` pass through. Any thoughts?